components/script/dom/readablestreamdefaultreader.rs

```diff
@@ -32,6 +39,199 @@ use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 pub enum ReadRequest {
     /// <https://streams.spec.whatwg.org/#default-reader-read>
     Read(Rc<Promise>),
+    /// <https://streams.spec.whatwg.org/#ref-for-read-request%E2%91%A2>
+    Tee { tee_read_request: TeeReadRequest },
+}
+
+#[derive(JSTraceable)]
+pub struct TeeReadRequest {
+    stream: Dom<ReadableStream>,
+    branch_1: Option<Dom<ReadableStream>>,
+    branch_2: Option<Dom<ReadableStream>>,
+    reading: Rc<Cell<bool>>,
+    read_again: Rc<Cell<bool>>,
+    canceled_1: Rc<Cell<bool>>,
+    canceled_2: Rc<Cell<bool>>,
+    clone_for_branch_2: Rc<Cell<bool>>,
+    cancel_promise: Rc<Promise>,
+    tee_underlying_source: Dom<TeeUnderlyingSource>,
+}
+
+impl TeeReadRequest {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        stream: Dom<ReadableStream>,
+        branch_1: Option<Dom<ReadableStream>>,
+        branch_2: Option<Dom<ReadableStream>>,
+        reading: Rc<Cell<bool>>,
+        read_again: Rc<Cell<bool>>,
+        canceled_1: Rc<Cell<bool>>,
+        canceled_2: Rc<Cell<bool>>,
+        clone_for_branch_2: Rc<Cell<bool>>,
+        cancel_promise: Rc<Promise>,
+        tee_underlying_source: Dom<TeeUnderlyingSource>,
+    ) -> Self {
+        TeeReadRequest {
+            stream,
+            branch_1,
+            branch_2,
+            reading,
+            read_again,
+            canceled_1,
+            canceled_2,
+            clone_for_branch_2,
+            cancel_promise,
+            tee_underlying_source,
+        }
+    }
+
+    /// Call into error of the default controller of branch_1,
+    /// <https://streams.spec.whatwg.org/#readable-stream-default-controller-error>
+    pub fn branch_1_default_controller_error(&self, error: SafeHandleValue) {
+        self.branch_1
+            .as_ref()
+            .expect("branch_1 must be set")
+            .get_default_controller()
+            .error(error);
+    }
+
+    /// Call into error of the default controller of branch_2,
+    /// <https://streams.spec.whatwg.org/#readable-stream-default-controller-error>
+    pub fn branch_2_default_controller_error(&self, error: SafeHandleValue) {
+        self.branch_2
+            .as_ref()
+            .expect("branch_2 must be set")
+            .get_default_controller()
+            .error(error);
+    }
+
+    /// Call into cancel of the stream,
+    /// <https://streams.spec.whatwg.org/#readable-stream-cancel>
+    pub fn stream_cancel(&self, reason: SafeHandleValue, can_gc: CanGc) {
+        self.stream.cancel(reason, can_gc);
+    }
+
+    /// <https://streams.spec.whatwg.org/#read-request-chunk-steps>
+    pub fn chunk_steps(&self, chunk: RootedTraceableBox<Heap<JSVal>>) {
+        // Set readAgain to false.
```

Comment:
So here you get the microtask queue with [this](https://github.com/servo/servo/blob/16da1c2721d471277c3981795d8d6000e8876cea/components/script/dom/globalscope.rs#L2990) and define a new [variant of micro task](https://github.com/servo/servo/blob/16da1c2721d471277c3981795d8d6000e8876cea/components/script/microtask.rs#L39) to run these steps and [enqueue it](https://github.com/servo/servo/blob/16da1c2721d471277c3981795d8d6000e8876cea/components/script/microtask.rs#L77). The micro trask works like a native promise handler, so you can store the requests and anything else you need on it and then run those steps using that data. I think you should define a method on the data inside your new microtask variant, and the method should run these steps when called into from the [micro task checkpoin](https://github.com/servo/servo/blob/16da1c2721d471277c3981795d8d6000e8876cea/components/script/microtask.rs#L113). The method could be called `microtask_chunk_steps`, and actually the inner data of the microtask could be this read request itself.